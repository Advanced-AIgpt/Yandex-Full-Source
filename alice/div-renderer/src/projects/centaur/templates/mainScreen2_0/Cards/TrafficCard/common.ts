import { ITrafficForecast } from './types';

export function sortTimeOfDay(originalArray: Readonly<ITrafficForecast[]>): ITrafficForecast[] {
    const forecastsFirst = [];
    const forecastsLast = [];

    for (const forecastElement of originalArray) {
        if (forecastElement.time === null || typeof forecastElement.time === 'undefined') {
            continue;
        }

        if (forecastsFirst.length === 0) {
            forecastsFirst.push(forecastElement);
        } else if (forecastsLast.length > 0) {
            forecastsLast.push(forecastElement);
        } else if (forecastElement.time - (forecastsFirst[forecastsFirst.length - 1].time as number) === 1) {
            forecastsFirst.push(forecastElement);
        } else {
            forecastsLast.push(forecastElement);
        }
    }

    return [...forecastsLast, ...forecastsFirst];
}
