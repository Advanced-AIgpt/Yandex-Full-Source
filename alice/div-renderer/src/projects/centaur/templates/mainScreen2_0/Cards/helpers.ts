export function getCardMainScreenId({ rowIndex, colIndex }: { rowIndex?: number; colIndex?: number }) {
    return `main_screen_card_${rowIndex || 0}_${colIndex || 0}`;
}
