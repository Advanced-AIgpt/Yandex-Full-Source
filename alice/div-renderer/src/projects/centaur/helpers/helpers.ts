/* eslint-disable camelcase */
import { sortedIndexBy } from 'lodash';
import { format, isSameDay } from 'date-fns';
import { ru } from 'date-fns/locale';
import { Div, IDivData, TextBlock } from 'divcard2';
import { IRequestState } from '../../../common/types/common';
import { getTemplates, templatesRegistry } from '../../../registries/common';
import { centaurTemplatesClass } from '../index';
import { Palette } from './palette';

export { default as pluralize } from '@yandex-int/i18n/lib/plural/ru';
type TextProps = Pick<ConstructorParameters<typeof TextBlock>[0], 'text_color'|'font_size'|'font_weight'|'line_height'>;

export const textBreakpoint =
    (...breakpoints: [number, TextProps][]) =>
        (text: string) => {
            const idx = Math.min(sortedIndexBy(breakpoints, [text.length, {}], 0), breakpoints.length - 1);
            return { text, ...breakpoints[idx][1] };
        };

export const formatDate = (d: Date | string | number, f: string) => format(new Date(d), f, { locale: ru });

export const textCommand = (c: string) => `dialog://text_command?query=${encodeURI(c)}`;

export const updatedAt = (d: Date) => {
    const now = new Date();
    return `Обновлен ${isSameDay(now, d) ? 'сегодня' : formatDate(d, 'dd.MM')}`;
};

export const insertBetween = <I extends unknown, S extends unknown>(itemList: I[], separator: S) => {
    return itemList.reduce<(I | S)[]>((result, item, index) => {
        if (index !== 0) {
            result.push(separator);
        }

        result.push(item);

        return result;
    }, []);
};

export interface ITemplateCard {
    templates: {
        [name: string]: Div;
    };
    palette?: Palette;
    card: IDivData;
}

// Дива верхнего уровня. Замена для new `TemplateCard()`
export const TopLevelCard = (card: Omit<IDivData, 'variable_triggers' | 'variables'>, requestState: IRequestState, palette?: Palette) => {
    const templates = Object.assign(
        getTemplates(requestState.localTemplates, templatesRegistry),
        centaurTemplatesClass.getUsedTemplates(requestState),
        requestState.res.localTemplates.getAll(),
    );

    return {
        templates,
        palette,
        card: {
            ...card,
            variable_triggers: requestState.variableTriggers.getAll(),
            variables: [...requestState.variables],
        } as IDivData,
    };
};
