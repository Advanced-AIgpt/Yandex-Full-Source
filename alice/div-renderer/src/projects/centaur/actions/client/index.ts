import { TemplateCard } from 'divcard2';
import { createClientAction } from '../../../../common/actions/client';
import { Layer } from '../../common/layers';
import { WebviewCard } from '../../components/cards';
import { ITemplateCard } from '../../helpers/helpers';

export enum EnumLayer {
    content = 'content',
    dialog = 'dialog',
}

export enum EnumInactivityTimeout {
    short = 'Short',
    medium = 'Medium',
    long = 'Long',
    infinity = 'Infinity',
}

export const localCommands = (commands: unknown[]) => {
    return `centaur://local_command?local_commands=[${commands.map(command => encodeURIComponent(JSON.stringify(command)))}]`;
};

export function createDirectiveToShowView({
    div2CardBody,
    doNotShowCloseButton = null,
    layer = EnumLayer.content,
    inactivityTimeout = EnumInactivityTimeout.infinity,
    actionSpaceId,
}: {
    div2CardBody: TemplateCard<string> | ITemplateCard,
    doNotShowCloseButton: boolean | null,
    layer: EnumLayer,
    inactivityTimeout: EnumInactivityTimeout,
    actionSpaceId: String | undefined,
}) {
    return {
        layer: {
            [layer]: {},
        },
        div2_card: {
            body: div2CardBody,
        },
        do_not_show_close_button: doNotShowCloseButton,
        inactivity_timeout: inactivityTimeout,
        action_space_id: actionSpaceId,
    };
}

export const localCommand = (command: unknown) => {
    return localCommands([command]);
};

export const closeLayerLocalAction = (layer = Layer.DIALOG, stopConversation = true) => {
    return localCommand({
        command: 'close_layer',
        layer: Layer[layer],
        stop_conversation: stopConversation,
    });
};

export const closeLayerAction = (logId: string, layer = Layer.DIALOG, stopConversation = true) => {
    return {
        log_id: logId,
        url: closeLayerLocalAction(layer, stopConversation),
    };
};


export const createAndroidIntentClientAction = (action: string, uri: string) => {
    return createClientAction('send_android_app_intent', {
        name: 'send_android_app_intent',
        action,
        uri,
    });
};

export const createClearQueueAction = () => {
    return createClientAction('clear_queue');
};

export const createWebviewClientAction = (webviewUrl: string) => {
    return createShowViewClientAction(WebviewCard(webviewUrl));
};

export const createShowViewClientAction = (
    div2CardBody: TemplateCard<string> | ITemplateCard,
    doNotShowCloseButton: boolean | null = null,
    layer = EnumLayer.content,
    inactivityTimeout = EnumInactivityTimeout.infinity,
    actionSpaceId: String | undefined = undefined,
) => {
    return createClientAction('show_view', createDirectiveToShowView({
        div2CardBody,
        doNotShowCloseButton,
        layer,
        inactivityTimeout,
        actionSpaceId,
    }));
};
