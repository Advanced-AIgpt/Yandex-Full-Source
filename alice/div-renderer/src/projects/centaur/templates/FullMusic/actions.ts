import { localCommands } from '../../actions/client';

export const ActionMusicScreenNext = (audioSourceId: string) => localCommands([
    {
        type: 'music_screen_next',
        audio_source_id: audioSourceId,
    },
]);

export const ActionMusicDislike = (audioSourceId: string) => localCommands([
    {
        type: 'audio_player_proxy_dislike',
        audio_source_id: audioSourceId,
    },
    'io_service_dislike',
    'io_service_next',
]);

export const ActionMusicLike = (audioSourceId: string) => localCommands([
    {
        type: 'audio_player_proxy_like',
        audio_source_id: audioSourceId,
    },
]);

export const ActionMusicExitFromIDLEMode = () => localCommands([
    'audio_player_proxy_exit_from_idle_mode',
]);

export const ActionMusicPlayPayse = () => localCommands([
    'io_service_toggle_play_pause',
]);

export const ActionMusicNext = () => localCommands([
    'audio_player_proxy_next',
    'io_service_next',
]);

export const ActionMusicPrev = () => localCommands([
    'audio_player_proxy_prev',
    'io_service_prev',
]);

export const ActionMusicShuffle = () => localCommands([
    'audio_player_proxy_shuffle',
]);

export const ActionMusicRepeat = () => localCommands([
    'audio_player_proxy_repeat',
]);
