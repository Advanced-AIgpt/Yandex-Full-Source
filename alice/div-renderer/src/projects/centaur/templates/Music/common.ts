import {
    ContainerBlock,
    Div,
    FixedSize,
    IDivAction,
    ImageBackground,
    ImageBlock, MatchParentSize,
    SolidBackground,
    TextBlock,
} from 'divcard2';
import { colorBlackOpacity70, offsetFromEdgeOfScreen } from '../../style/constants';
import { compact } from 'lodash';
import EmptyDiv from '../../components/EmptyDiv';
import { text28m } from '../../style/Text/Text';
import { directivesAction } from '../../../../common/actions';
import { centaurMusicPlay } from '../../../../common/actions/server/musicActions';
import MusicPlaying from './MusicPlaying/MusicPlaying';
import { IRequestState } from '../../../../common/types/common';
import LikeButton from './components/LikeButton';

interface IGetPlaylistItemProps {
    title: string;
    subtitle?: string;
    objectId: string;
    objectType: string;
    startFromTrackId?: string;
    isCurrent?: boolean;
    imageUrl: string;
    imageBackground?: string;
    requestState: IRequestState;
    isLiked?: boolean;
}

function CoverWithBackground(imageUrl: string, background: string, action: IDivAction, isCurrent: boolean, requestState: IRequestState) {
    return new ContainerBlock({
        width: new FixedSize({ value: 80 }),
        height: new FixedSize({ value: 80 }),
        alignment_vertical: 'top',
        actions: [action],
        margins: {
            right: 24,
        },
        border: {
            corner_radius: 40,
        },
        content_alignment_horizontal: 'center',
        content_alignment_vertical: 'center',
        background: compact([
            new SolidBackground({
                color: background,
            }),
        ]),
        orientation: 'overlap',
        items: compact([
            new ImageBlock({
                image_url: imageUrl,
                width: new FixedSize({ value: 48 }),
                height: new FixedSize({ value: 48 }),
            }),
            isCurrent && new EmptyDiv({
                width: new FixedSize({ value: 80 }),
                height: new FixedSize({ value: 80 }),
                background: [
                    new SolidBackground({
                        color: colorBlackOpacity70,
                    }),
                ],
            }),
            isCurrent && MusicPlaying(requestState),
        ]),
    });
}

function CoverWithoutBackground(imageUrl: string, action: IDivAction, isCurrent: boolean, requestState: IRequestState) {
    return new ContainerBlock({
        width: new FixedSize({ value: 80 }),
        height: new FixedSize({ value: 80 }),
        alignment_vertical: 'top',
        content_alignment_horizontal: 'center',
        content_alignment_vertical: 'center',
        actions: [action],
        margins: {
            right: 24,
        },
        border: {
            corner_radius: 20,
        },
        background: compact([
            new ImageBackground({
                image_url: imageUrl,
            }),
            isCurrent && new SolidBackground({
                color: colorBlackOpacity70,
            }),
        ]),
        items: [
            isCurrent ? MusicPlaying(requestState) : new EmptyDiv(),
        ],
    });
}

export function getPlaylistItem({
    isCurrent = false,
    objectType,
    subtitle,
    objectId,
    title,
    startFromTrackId,
    imageUrl,
    imageBackground,
    requestState,
    isLiked,
}: IGetPlaylistItemProps): {
    options: Partial<ConstructorParameters<typeof ContainerBlock>[0]>;
    items: Div[];
} {
    const openAction = {
        log_id: 'play_music_click',
        url: directivesAction(centaurMusicPlay({
            disableNlg: true,
            objectType,
            objectId,
            startFromTrackId,
        })),
    };

    return {
        options: {
            orientation: 'horizontal',
            paddings: {
                top: 28,
                left: offsetFromEdgeOfScreen,
                bottom: 28,
                right: offsetFromEdgeOfScreen,
            },
            content_alignment_vertical: 'center',
        },
        items: compact([
            imageBackground ?
                CoverWithBackground(imageUrl, imageBackground, openAction, isCurrent, requestState) :
                CoverWithoutBackground(imageUrl, openAction, isCurrent, requestState),
            new ContainerBlock({
                width: new MatchParentSize({ weight: 1 }),
                actions: [openAction],
                items: compact([
                    new TextBlock({
                        ...text28m,
                        text: title,
                    }),
                    subtitle && new TextBlock({
                        ...text28m,
                        text: subtitle,
                        alpha: 0.5,
                    }),
                ]),
            }),
            typeof isLiked !== 'undefined' && LikeButton({ isLiked, trackId: startFromTrackId || '' }),
        ]),
    };
}
