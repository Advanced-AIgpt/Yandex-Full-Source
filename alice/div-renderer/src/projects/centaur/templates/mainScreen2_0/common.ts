import { IDivAction } from 'divcard2';
import { EnumLayer } from '../../actions/client';
import { CrutchLoader } from '../../components/Loader/Loader';

interface IPutLoaderToActionOptions {
    action?: string | null,
    layer: EnumLayer,
    actionLogId?: string,
}

export const putLoaderToAction = ({
    action,
    layer,
    actionLogId = 'action_with_loader',
}: IPutLoaderToActionOptions): IDivAction[] | undefined => {
    if (!action) {
        return undefined;
    }
    return [
        CrutchLoader(layer),
        {
            log_id: actionLogId,
            url: action,
        },
    ];
};
