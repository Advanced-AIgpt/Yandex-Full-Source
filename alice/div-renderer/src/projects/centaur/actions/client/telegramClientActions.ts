import { createShowViewClientAction, EnumInactivityTimeout, EnumLayer } from './index';
import { TopLevelCard } from '../../helpers/helpers';
import { BoolIntVariable, ContainerBlock, DivVariable, MatchParentSize } from 'divcard2';
import { createClientAction, IDirectiveClientAction } from '../../../../common/actions/client';
import { TelegramLogin } from '../../divCustoms/TelegramCustoms';
import { createSemanticFrameAction } from '../../../../common/actions/server/index';
import { IRequestState } from '../../../../common/types/common';
import { createRequestState } from '../../../../registries/common';
import { CurrentCallCard, ICurrentCallData } from '../../templates/telegram/CurrentCall';
import { MMRequest } from '../../../../common/helpers/MMRequest';

export const bindToCallVariable: DivVariable = new BoolIntVariable({
    name: 'bindToCall',
    value: 1,
});

export function telegramDiscardVideoCallDirective(
    userId: string,
    callId: string | undefined = undefined,
): IDirectiveClientAction {
    return createClientAction(
        'discard_video_call_directive',
        {
            telegram_discard_video_call_data: {
                call_owner_data: {
                    user_id: userId,
                    call_id: callId,
                },
            },
        },
    );
}

export function telegramAcceptVideoCallDirective(
    callId: string,
    userId: string,
): IDirectiveClientAction {
    return createClientAction(
        'accept_video_call_directive',
        {
            telegram_accept_video_call_data: {
                call_owner_data: {
                    call_id: callId,
                    user_id: userId,
                },
            },
        },
    );
}

export function telegramAcceptVideoCallShowView(
    mmRequest: MMRequest,
    currentData: ICurrentCallData,
): IDirectiveClientAction {
    const actionRequestState = createRequestState(mmRequest);

    return createShowViewClientAction(
        CurrentCallCard(currentData, actionRequestState),
        true,
        EnumLayer.content,
        EnumInactivityTimeout.short,
        'video_call.current_call.action_space_id',
    );
}

export function telegramStartLoginDirective(
    id: string,
): IDirectiveClientAction {
    return createClientAction(
        'start_video_call_login_directive',
        {
            telegram_start_login_data: {
                id,
                on_fail_callback: createSemanticFrameAction(
                    {
                        video_call_login_failed_semantic_frame: {
                            provider: {
                                enum_value: 'Telegram',
                            },
                        },
                    },
                    'VideoCall',
                    'video_call_login_failed',
                ),
            },
        },
    );
}

export function telegramMuteMicDirective(
    userId: string,
    callId: string | undefined = undefined,
): IDirectiveClientAction {
    return createClientAction(
        'video_call_mute_mic_directive',
        {
            telegram_mute_mic_data: {
                call_owner_data: {
                    call_id: callId,
                    user_id: userId,
                },
            },
        },
    );
}

export function telegramUnmuteMicDirective(
    userId: string,
    callId: string | undefined = undefined,
): IDirectiveClientAction {
    return createClientAction(
        'video_call_unmute_mic_directive',
        {
            telegram_unmute_mic_data: {
                call_owner_data: {
                    call_id: callId,
                    user_id: userId,
                },
            },
        },
    );
}

export function turnOnVideoDirective(
    userId: string,
    callId: string | undefined = undefined,
): IDirectiveClientAction {
    return createClientAction(
        'video_call_turn_on_video_directive',
        {
            telegram_turn_on_video_data: {
                call_owner_data: {
                    call_id: callId,
                    user_id: userId,
                },
            },
        },
    );
}

export function turnOffVideoDirective(
    userId: string,
    callId: string | undefined = undefined,
): IDirectiveClientAction {
    return createClientAction(
        'video_call_turn_off_video_directive',
        {
            telegram_turn_off_video_data: {
                call_owner_data: {
                    call_id: callId,
                    user_id: userId,
                },
            },
        },
    );
}

export function telegramStartLoginShowView(
    id: string,
    requestState: IRequestState,
): IDirectiveClientAction {
    return createShowViewClientAction(
        TopLevelCard({
            log_id: 'telegram_login_screen',
            states: [
                {
                    state_id: 0,
                    div: new ContainerBlock({
                        width: new MatchParentSize(),
                        height: new MatchParentSize(),
                        orientation: 'overlap',
                        items: [
                            TelegramLogin(id),
                        ],
                    }),
                },
            ],
        }, requestState),
        false,
        EnumLayer.dialog,
        EnumInactivityTimeout.infinity,
    );
}

