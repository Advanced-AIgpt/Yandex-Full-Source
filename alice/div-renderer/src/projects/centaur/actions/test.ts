import { ContainerBlock, DivCustomBlock, MatchParentSize } from 'divcard2';
import { createWebviewClientAction } from './client';
import { AnonymizeDataForSnapshot } from '../../../common/helpers/dev';
import { directivesAction } from '../../../common/actions';

describe('Client action directive', () => {
    it('matches snapshot', () => {
        expect(AnonymizeDataForSnapshot(directivesAction([{
            type: 'client_action', name: 'show_view', payload: {
                layer: {
                    content: {},
                },
                div2_card: {
                    body: {
                        card: {
                            states: [
                                {
                                    state_id: 0,
                                    div: new ContainerBlock({
                                        items: [
                                            new DivCustomBlock({
                                                custom_type: 'centaur_webview',
                                                custom_props: {
                                                    url: 'https://yandex.ru/quasar',
                                                },
                                                width: new MatchParentSize(),
                                                height: new MatchParentSize(),
                                                alignment_vertical: 'bottom',
                                            }),
                                        ],
                                        width: new MatchParentSize(),
                                        height: new MatchParentSize(),
                                        alignment_vertical: 'bottom',
                                    }),
                                },
                            ],
                            log_id: 'iot-web',
                        },
                        templates: {},
                    },
                },
            },
        }]))).toMatchSnapshot();
    });
});

describe('Create webview client action', () => {
    it('matches snapshot', () => {
        expect(AnonymizeDataForSnapshot(createWebviewClientAction('https://yandex.ru'))).toMatchSnapshot();
    });
});
