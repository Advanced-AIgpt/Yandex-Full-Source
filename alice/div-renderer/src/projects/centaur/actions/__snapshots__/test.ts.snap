// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Client action directive matches snapshot 1`] = `"dialog://?directives=%5B%7B%22type%22%3A%22client_action%22%2C%22name%22%3A%22show_view%22%2C%22payload%22%3A%7B%22layer%22%3A%7B%22content%22%3A%7B%7D%7D%2C%22div2_card%22%3A%7B%22body%22%3A%7B%22card%22%3A%7B%22states%22%3A%5B%7B%22state_id%22%3A0%2C%22div%22%3A%7B%22type%22%3A%22container%22%2C%22alignment_vertical%22%3A%22bottom%22%2C%22height%22%3A%7B%22type%22%3A%22match_parent%22%7D%2C%22items%22%3A%5B%7B%22type%22%3A%22custom%22%2C%22alignment_vertical%22%3A%22bottom%22%2C%22custom_props%22%3A%7B%22url%22%3A%22https%3A%2F%2Fyandex.ru%2Fquasar%22%7D%2C%22custom_type%22%3A%22centaur_webview%22%2C%22height%22%3A%7B%22type%22%3A%22match_parent%22%7D%2C%22width%22%3A%7B%22type%22%3A%22match_parent%22%7D%7D%5D%2C%22width%22%3A%7B%22type%22%3A%22match_parent%22%7D%7D%7D%5D%2C%22log_id%22%3A%22iot-web%22%7D%2C%22templates%22%3A%7B%7D%7D%7D%7D%7D%5D"`;

exports[`Create webview client action matches snapshot 1`] = `
Object {
  "name": "show_view",
  "payload": Object {
    "div2_card": Object {
      "body": Object {
        "card": Object {
          "log_id": "webview",
          "states": Array [
            Object {
              "div": Object {
                "height": Object {
                  "type": "match_parent",
                },
                "items": Array [
                  Object {
                    "alignment_vertical": "bottom",
                    "custom_props": Object {
                      "url": "https://yandex.ru",
                      "user_agent": "Mozilla/5.0 (Linux; Android 9; SM-A102U) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/74.0.3729.136 Mobile Safari/537.36",
                    },
                    "custom_type": "centaur_webview",
                    "height": Object {
                      "type": "match_parent",
                    },
                    "type": "custom",
                    "width": Object {
                      "type": "match_parent",
                    },
                  },
                  Object {
                    "alignment_vertical": "top",
                    "height": Object {
                      "type": "wrap_content",
                    },
                    "items": Array [
                      Object {
                        "alignment_horizontal": "right",
                        "alignment_vertical": "top",
                        "height": Object {
                          "type": "wrap_content",
                        },
                        "items": Array [
                          Object {
                            "action": Object {
                              "log_id": "close_webview",
                              "url": "centaur://local_command?local_commands=[{'command':'close_layer','layer':'CONTENT'}]",
                            },
                            "height": Object {
                              "type": "fixed",
                              "value": 56,
                            },
                            "image_url": "https://yastatic.net/s3/dialogs/smart_displays/icons/close_wt_border.png",
                            "preload_required": 1,
                            "type": "image",
                            "width": Object {
                              "type": "fixed",
                              "value": 56,
                            },
                          },
                        ],
                        "margins": Object {
                          "bottom": 16,
                          "right": 24,
                          "top": 16,
                        },
                        "orientation": "horizontal",
                        "type": "container",
                        "width": Object {
                          "type": "wrap_content",
                        },
                      },
                      Object {
                        "alignment_horizontal": "left",
                        "alignment_vertical": "top",
                        "height": Object {
                          "type": "wrap_content",
                        },
                        "items": Array [
                          Object {
                            "action": Object {
                              "log_id": "webview_go_back",
                              "url": "centaur://local_command?local_commands=[{'command':'webview_go_back','layer':'CONTENT'}]",
                            },
                            "height": Object {
                              "type": "fixed",
                              "value": 56,
                            },
                            "image_url": "https://yastatic.net/s3/dialogs/smart_displays/icons/back_wt_border.png",
                            "preload_required": 1,
                            "type": "image",
                            "width": Object {
                              "type": "fixed",
                              "value": 56,
                            },
                          },
                        ],
                        "margins": Object {
                          "bottom": 16,
                          "left": 24,
                          "top": 16,
                        },
                        "orientation": "horizontal",
                        "type": "container",
                        "width": Object {
                          "type": "wrap_content",
                        },
                      },
                    ],
                    "orientation": "overlap",
                    "type": "container",
                    "width": Object {
                      "type": "match_parent",
                    },
                  },
                ],
                "orientation": "vertical",
                "type": "container",
                "width": Object {
                  "type": "match_parent",
                },
              },
              "state_id": 0,
            },
          ],
        },
        "templates": Object {},
      },
    },
    "do_not_show_close_button": null,
    "inactivity_timeout": "Infinity",
    "layer": Object {
      "content": Object {},
    },
  },
  "type": "client_action",
}
`;
